pkg/client/unversioned/clientcmd/merged_client_builder.go

```diff
@@ -112,8 +112,11 @@ func (config *DeferredLoadingClientConfig) ClientConfig() (*restclient.Config, e
 		//   "empty due to defaults"
 		// TODO: this shouldn't be a global - the client config rules should be
 		//   handling this.
-		defaultConfig, err := DefaultClientConfig.ClientConfig()
-		if err == nil && !reflect.DeepEqual(mergedConfig, defaultConfig) {
+		defaultConfig, defErr := DefaultClientConfig.ClientConfig()
+		if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
```

Comment:
Let's list the cases that should return the original base config:

1. base config was returned and != defaultConfig
2. default config is invalid and merged config was not empty

The cases that should return errors:

1. base config returned any error except IsEmpty

The cases that should fall through to in-cluster are 

1. base config is valid and non-default
2. base config is empty AND (default config is empty OR default config == base config)

It looks like the code you have here does that after deeper inspection, but I want to make sure we're all on the same page with the descriptions.